### PR TITLE
Adds support for XD84 Pro

### PR DESCRIPTION
**Pull request by katawajojo**
Adds QMK support for KPRepublic's XD84 Pro PCB.
Includes ANSI, ISO, and VIA (ANSI layout only currently) Support.

## Description

Adds support for KPRepublic's XD84 Pro PCB.
The PCB's silkscreen refers to this board as "XD84 v2.0"

I've also created a VIA pull request as well.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/xd84pro/info.json [hunk 1]
```diff
@@ -0,0 +1,188 @@
+{
+    "keyboard_name": "XD84 Pro",
+    "url": "",
+    "maintainer": "qmk",
+    "width": 16,
+    "height": 6,
+    "layouts": {
+        "LAYOUT_75_ansi": {
+            "key_count": 84,
```
Comment (fauxpark) [inline keyboards/xd84pro/info.json:9]:
```suggestion
```

File: keyboards/xd84pro/info.json [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+                {"label":"Alt", "x":10, "y":5},
+                {"label":"Fn", "x":11, "y":5},
+                {"label":"Ctrl", "x":12, "y":5},
+                {"label":"Left", "x":13, "y":5},
+                {"label":"Down", "x":14, "y":5},
+                {"label":"Right", "x":15, "y":5}
+            ]
+        },
+        "LAYOUT_75_iso": {
+            "key_count": 85,
```
Comment (fauxpark) [inline keyboards/xd84pro/info.json:98]:
```suggestion
```

File: keyboards/xd84pro/keymaps/via/rules.mk [hunk 3]
```diff
@@ -0,0 +1,2 @@
+LINK_TIME_OPTIMIZATION_ENABLE = yes
+VIA_ENABLE = yes
```
Comment (fauxpark) [inline keyboards/xd84pro/keymaps/via/rules.mk:2]:
```suggestion
VIA_ENABLE = yes
LTO_ENABLE = yes
```

File: keyboards/xd84pro/readme.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+Keyboard with 84 Keys, RGB LED Underglow, and USB-C Connection
+- Designed by Xiudi
+- Up to 87 keys
+  - iso and ansi support
+
+The silkscreen printing on the PCB refers to this as the "XD84 v2.0".
+
+Keyboard Maintainer: QMK Community
+Hardware Supported: XD84 Pro
+Hardware Availability: [KPrepublic](https://kprepublic.com/products/xd84pro-xd84-pro-custom-mechanical-keyboard-kit-75-supports-tkg-tools-support-underglow-rgb-pcb-programmed-gh84-kle-type-c)
```
Comment (fauxpark) [inline keyboards/xd84pro/readme.md:14]:
```suggestion
* Keyboard Maintainer: QMK Community
* Hardware Supported: XD84 Pro
* Hardware Availability: [KPrepublic](https://kprepublic.com/products/xd84pro-xd84-pro-custom-mechanical-keyboard-kit-75-supports-tkg-tools-support-underglow-rgb-pcb-programmed-gh84-kle-type-c)
```

File: keyboards/xd84pro/readme.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+Make example for this keyboard (after setting up your build environment):
+
+    make xd84pro:default
+
+Install examples:
+
+    make xd84pro:default:dfu
```
Comment (fauxpark) [inline keyboards/xd84pro/readme.md:22]:
```suggestion
Flashing example for this keyboard:

    make xd84pro:default:flash
```

File: keyboards/xd84pro/rules.mk [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
Comment (fauxpark) [inline keyboards/xd84pro/rules.mk:11]:
```suggestion
```

File: keyboards/xd84pro/rules.mk [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs
```
Comment (fauxpark) [inline keyboards/xd84pro/rules.mk:33]:
```suggestion
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/xd84pro/xd84pro.c [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+void led_set_kb(uint8_t usb_led) {
+  if (IS_LED_ON(usb_led, USB_LED_CAPS_LOCK)) {
+    writePinLow(B2);
+  } else {
+    writePinHigh(B2);
+  }
+
+  led_set_user(usb_led);
+}
```
Comment (fauxpark) [inline keyboards/xd84pro/xd84pro.c:32]:
```suggestion
void keyboard_pre_init_kb(void) {
    led_init_ports();
    keyboard_pre_init_user();
}

void led_init_ports(void) {
    setPinOutput(B2);
}

bool led_update_kb(led_t led_state) {
    if (led_update_user(led_state)) {
        writePin(B2, !led_state.caps_lock);
    }

    return false;
}
```

**Comment by katawajojo** [comment]
Thanks for the suggestions, I’ll take care of them when I wake up in the morning.

**Comment by sigprof** [comment]
You should make a `LAYOUT_all` macro (and the corresponding entry in info.json) which includes all possible key positions. At the moment there are two missing parts:

1. According to the vendor documentation ([1](https://ae01.alicdn.com/kf/H6aa7f4160b55408facb8bde2f64a31bcI.jpg), [2](https://ae01.alicdn.com/kf/H9b653564abec4419825214f93ffc675aZ.jpg)), this PCB supports splitting the 2U Backspace key into 2×1U keys. You should figure out the matrix positions of those keys (one of them probably occupies some of remaining positions on the last row, the other is most likely the same as the 2U Backspace) and add them to `LAYOUT_all`.

2. This PCB, like other XD* PCBs, not just has three distinct matrix positions for the ANSI backslash, ISO NUHS and Enter, but actually supports installing 1U+1.25U keys in the place normally occupied by the 2.25U ANSI Enter. Again, you should check how those matrix positions are connected (although in this case the fact that the 1.25U part of ANSI Enter is connected in parallel to the normal ANSI Enter pads is even visible [on this shop photo](https://ae01.alicdn.com/kf/H4456b675111a4588a6e1f9e83bfc5b3em.jpg)), and add all those 3 keys to `LAYOUT_all`.

The rest of `LAYOUT_all` should be the same as `LAYOUT_75_iso` (so that the NUBS key would be present).

After adding this `LAYOUT_all` macro you should use it in the `via` keymap to make a layout which could work for both ISO and ANSI out of the box (this should be possible on this PCB, even the ANSI backslash and ISO NUHS could be made distinct). For the split Backspace you could just assign `KC_BSPC` to both halves of the 2U Backspace.

**Comment by Popkornium18** [comment]
@katawajojo are you still planning to make the requested changes? If not I might give it a go, since I picked up this board thinking it was supported :facepalm:

**Comment by katawajojo** [comment]
Sorry! I've been working on these changes, but I've also been hella busy with school coming up and work. Give me like, a day or two to finish these up, and update the pull.

**Comment by katawajojo** [comment]
Okay, so I'm going to remove the VIA stuff for now, as I need to do some research on doing that all properly. I have every layout mapped and ready, I'm just not sure how to go about properly creating the `LAYOUT_all` macro.

Side note: this keyboard has a backlighting issue that's resolved with the pull request #9575. Applying the code from this pull request fixes the flickering backlight.

Attached is the matrix layout for all keys, and all supported configurations of the board.

**Comment by Popkornium18** [comment]
> Side note: this keyboard has a backlighting issue that's resolved with the pull request #9575. Applying the code from this pull request fixes the flickering backlight.

It doesn't fix the problem for me :thinking:

**Comment by katawajojo** [comment]
> > Side note: this keyboard has a backlighting issue that's resolved with the pull request #9575. Applying the code from this pull request fixes the flickering backlight.
> 
> 
> 
> It doesn't fix the problem for me :thinking: 

Odd, as this fixed the flickering issue on both XD84 Pro boards I have.

**Review by drashna — APPROVED** [review]

**Comment by Popkornium18** [comment]
> Side thought: Do you have the RGB underglow cycling through effects at all? or is it just a solid color and brightness?
No I was using the cycling effect. With the underglow disabled the flickering goes away. :thinking:

**Comment by katawajojo** [comment]
> > Side thought: Do you have the RGB underglow cycling through effects at all? or is it just a solid color and brightness?
> 
> 
> 
> No I was using the cycling effect. With the underglow disabled the flickering goes away. :thinking: 

Okay, yeah, underglow effects can still cause issues with backlight flickering. If you read through that pull request, it mentions this. Iirc, this is just a bug within QMK itself.

File: keyboards/xd84pro/rules.mk [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
```
Comment (fauxpark) [inline keyboards/xd84pro/rules.mk:22]:
```suggestion
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/xd84pro/config.h [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0xCDCD
+#define PRODUCT_ID      0x8484
```
[the remark this answers is not in the document]
Comment (katawajojo) [reply keyboards/xd84pro/config.h:23]: I used the original IDs for this board. Is there a reason for this change? Just out of curiosity.

[the remark this answers is not in the document]
Comment (fauxpark) [reply keyboards/xd84pro/config.h:23]: It matches the other XD boards.

**Comment by katawajojo** [comment]
There we go, accepted those changes. I'll also have to change that info for the VIA request as well, to match the vendor ID and the product ID, however.

**Review by fauxpark — CHANGES_REQUESTED** [review]
Looks good, just waiting on the USB ID changes and VIA keymap.

File: keyboards/xd84pro/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,36 @@
+#include QMK_KEYBOARD_H
```
Comment (noroadsleft) [inline keyboards/xd84pro/keymaps/default/keymap.c:1]: Could we get a license header here?

Here's a sample one:

https://github.com/qmk/qmk_firmware/blob/4a83021fe0d07888e0e3bf161882c4ab7221a4a6/quantum/template/base/keymaps/default/keymap.c#L1-L15

File: keyboards/xd84pro/keymaps/default_iso/keymap.c [hunk 12]
```diff
@@ -0,0 +1,38 @@
+#include QMK_KEYBOARD_H
```
Comment (noroadsleft) [inline keyboards/xd84pro/keymaps/default_iso/keymap.c:1]: And a license header here too.

**Review by noroadsleft — CHANGES_REQUESTED** [review]

**Comment by fauxpark** [comment]
@katawajojo any updates on this? We're still missing the VIA keymap, but other than that and noroads' review above, I think this is good to go.

**Comment by katawajojo** [comment]
I've been extra busy lately with work and school. I'm working on trying to make that VIA keymap, and adding the extra bottom row configurations. Thankfully, I have nothing but free time today, so I should be able to knock everything out.

**Comment by katawajojo** [comment]
Okay, I've got everything ready, but I'm not 100% sure on making the new VIA .json file. If anyone can help out with that, that would be SUPER helpful. However, for now, everything should be ready to go. I'm gonna test before pushing.

Once again, sorry for the extremely long wait time, life's gotten extra busy on me lately.

**Comment by katawajojo** [comment]
Okay, this should be everything. I'm getting the VIA json cleaned up for the pull request over there as well, but this should be everything needed for over here. I've tested this with my own PCB and everything seems to check out fine. I'm not sure what's failing in the Lint though.

File: keyboards/xd84pro/info.json [hunk 13]
```diff
@@ -0,0 +1,280 @@
+{
+    "keyboard_name": "XD84 Pro",
+    "url": "",
+    "maintainer": "qmk",
+    "width": 16,
+    "height": 6,
+    "layouts": {
+        "LAYOUT_75_all": {
+            "key_count": 87,
```
Comment (fauxpark) [inline keyboards/xd84pro/info.json:9]:
```suggestion
```

File: keyboards/xd84pro/rules.mk [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = yes           # USB Nkey Rollover
+BACKLIGHT_ENABLE = yes      # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = yes       # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
+
+LAYOUTS = 75_all 75_ansi 75_iso
```
Comment (fauxpark) [inline keyboards/xd84pro/rules.mk:24]:
```suggestion
LAYOUTS = 75_ansi 75_iso
```

File: keyboards/xd84pro/xd84pro.h [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+/* This a shortcut to help you visually see your layout.
+ *
+ * The first section contains all of the arguments representing the physical
+ * layout of the board and position of the keys.
+ *
+ * The second converts the arguments into a two-dimensional array which
+ * represents the switch matrix.
+ */
+#define LAYOUT_75_all( \
```
Comment (fauxpark) [inline keyboards/xd84pro/xd84pro.h:29]:
```suggestion
#define LAYOUT_all( \
```

File: keyboards/xd84pro/keymaps/via/keymap.c [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+  *
+  * You should have received a copy of the GNU General Public License
+  * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+  */
+
+#include QMK_KEYBOARD_H
+
+const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
+
+  [0] = LAYOUT_75_all(
```
Comment (fauxpark) [inline keyboards/xd84pro/keymaps/via/keymap.c:21]:
```suggestion
  [0] = LAYOUT_all(
```

File: keyboards/xd84pro/keymaps/via/keymap.c [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+      KC_CAPS,          KC_A,     KC_S,     KC_D,     KC_F,     KC_G,     KC_H,     KC_J,     KC_K,     KC_L,     KC_SCLN,  KC_QUOT,  KC_BSLS,   KC_ENT,   KC_PGDN,
+/* ├─────────────────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───────────┼─────────┼─────────┤ */
+      KC_LSFT, KC_LSFT,     KC_Z,     KC_X,     KC_C,     KC_V,     KC_B,     KC_N,     KC_M,     KC_COMM,  KC_DOT,   KC_SLSH,  KC_RSFT,          KC_UP,    KC_END,
+/* ├───────────┬─────────┴─┬───────┴───┬─────┴─────────┴─────────┴─────────┴─────────┴─────────┴───────┬─┴───────┬─┴───────┬─┴───────┬─────────┼─────────┼─────────┤ */
+      KC_LCTL,    KC_LGUI,    KC_LALT,    KC_SPC,                                                         KC_RALT,    MO(1),  KC_RCTL,  KC_LEFT,  KC_DOWN,  KC_RGHT
+/* └───────────┴───────────┴───────────┴───────────────────────────────────────────────────────────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘ */
+  ),
+
+  [1] = LAYOUT_75_all(
```
Comment (fauxpark) [inline keyboards/xd84pro/keymaps/via/keymap.c:37]:
```suggestion
  [1] = LAYOUT_all(
```

File: keyboards/xd84pro/keymaps/via/keymap.c [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+      _______,          _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,    _______,
+/* ├─────────────────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───────────┬─────────┼─────────┤ */
+      _______,  _______,    _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,          _______,  _______,
+/* ├───────────┬─────────┴─┬───────┴───┬─────┴─────────┴─────────┴─────────┴─────────┴─────────┴───────┬─┴───────┬─┴───────┬─┴───────┬─────────┼─────────┼─────────┤ */
+      _______,    _______,    _______,    _______,                                                        _______,  _______,  _______,  _______,  _______,  _______
+/* └───────────┴───────────┴───────────┴───────────────────────────────────────────────────────────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘ */
+  ),
+
+  [2] = LAYOUT_75_all(
```
Comment (fauxpark) [inline keyboards/xd84pro/keymaps/via/keymap.c:53]:
```suggestion
  [2] = LAYOUT_all(
```

File: keyboards/xd84pro/keymaps/via/keymap.c [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+      _______,          _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,    _______,
+/* ├─────────────────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───┬─────┴───────────┬─────────┼─────────┤ */
+      _______,  _______,    _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,  _______,          _______,  _______,
+/* ├───────────┬─────────┴─┬───────┴───┬─────┴─────────┴─────────┴─────────┴─────────┴─────────┴───────┬─┴───────┬─┴───────┬─┴───────┬─────────┼─────────┼─────────┤ */
+      _______,    _______,    _______,    _______,                                                        _______,  _______,  _______,  _______,  _______,  _______
+/* └───────────┴───────────┴───────────┴───────────────────────────────────────────────────────────────┴─────────┴─────────┴─────────┴─────────┴─────────┴─────────┘ */
+  ),
+
+  [3] = LAYOUT_75_all(
```
Comment (fauxpark) [inline keyboards/xd84pro/keymaps/via/keymap.c:69]:
```suggestion
  [3] = LAYOUT_all(
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/xd84pro/info.json [hunk 20]
```diff
@@ -0,0 +1,279 @@
...
+                {"label":"Alt", "x":10, "y":5},
+                {"label":"Fn", "x":11, "y":5},
+                {"label":"Ctrl", "x":12, "y":5},
+                {"label":"Left", "x":13, "y":5},
+                {"label":"Down", "x":14, "y":5},
+                {"label":"Right", "x":15, "y":5}
+            ]
+        },
+        "LAYOUT_75_ansi": {
+            "key_count": 84,
```
Comment (fauxpark) [inline keyboards/xd84pro/info.json:100]:
```suggestion
```

File: keyboards/xd84pro/info.json [hunk 21]
```diff
@@ -0,0 +1,279 @@
...
+                {"label":"Alt", "x":10, "y":5},
+                {"label":"Fn", "x":11, "y":5},
+                {"label":"Ctrl", "x":12, "y":5},
+                {"label":"Left", "x":13, "y":5},
+                {"label":"Down", "x":14, "y":5},
+                {"label":"Right", "x":15, "y":5}
+            ]
+        },
+        "LAYOUT_75_iso": {
+            "key_count": 85,
```
Comment (fauxpark) [inline keyboards/xd84pro/info.json:189]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by katawajojo** [comment]
For future knowledge, should I forgo using "key_count" in the info.json?

**Comment by fauxpark** [comment]
Yep. It's not really necessary as you can just count the number of elements in the `layout` array.

**Comment by katawajojo** [comment]
> Yep. It's not really necessary as you can just count the number of elements in the `layout` array.

Huh! I never thought about that. Good to know, thanks! c:

File: keyboards/xd84pro/info.json [hunk 22]
```diff
@@ -0,0 +1,277 @@
+{
+    "keyboard_name": "XD84 Pro",
+    "url": "",
+    "maintainer": "qmk",
+    "width": 16,
+    "height": 6,
+    "layouts": {
+        "LAYOUT_75_all": {
```
Comment (fauxpark) [inline keyboards/xd84pro/info.json:8]:
```suggestion
        "LAYOUT_all": {
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
Whoops, one last thing then I think this is good to go.

**Review by fauxpark — APPROVED** [review]

**Comment by katawajojo** [comment]
I also just updated the VIA .json file and updated the VIA pull request as well. This should complete the VIA and QMK port for this PCB.

**Comment by katawajojo** [comment]
Any update on this?

**Comment by fauxpark** [comment]
Thanks!

**Pull request merged**